services/QuillLMS/app/services/skills_report/question_responses_base.rb

```diff
@@ -0,0 +1,104 @@
+    end
+
+    def run = raw_results.map { |h| self.class::Result.new(**h) }
+
+    private def raw_results = RawSqlRunner.execute(sql)
+
+    private def sql = ActiveRecord::Base.sanitize_sql([query_sql, { classroom_id:, concept_ids: }])
+
+    private def query_sql
+      <<-SQL
+        #{common_ctes}
+        #{select_sql}
+      SQL
+    end
+
+    private def common_ctes
+      <<-SQL
+        WITH concept_uids AS (
+            SELECT c.uid
+            FROM concepts c
+            LEFT JOIN concepts parent_c on c.parent_id = parent_c.id
+            LEFT JOIN concepts grandparent_c on parent_c.parent_id = grandparent_c.id
+            WHERE (c.id IN (:concept_ids) OR c.parent_id IN (:concept_ids) OR parent_c.parent_id IN (:concept_ids))
+            #{ell_where_clause}
+        ),
+        relevant_sessions AS MATERIALIZED (
+            SELECT
+              activity_sessions.id,
+              user_id,
+              activity_id,
+              timespent,
+              classroom_units.classroom_id
+            FROM activity_sessions
+            JOIN classroom_units ON activity_sessions.classroom_unit_id = classroom_units.id
+            JOIN students_classrooms ON activity_sessions.user_id = students_classrooms.student_id
+              AND students_classrooms.visible = true
+            WHERE students_classrooms.classroom_id = :classroom_id
+              AND activity_sessions.completed_at IS NOT NULL
+              #{session_classroom_where_clause}
+        ),
+        relevant_concept_results AS (
+            SELECT
+                    /* With the exception of attempt_number, all of these MAX calls are against columns that should have the same value in all aggregated rows, so logically we're not expecting a difference between using MAX or MIN here */
+                    cr.activity_session_id,
+                    cr.question_number,
+                    MAX(cr.extra_metadata->>'question_uid') AS question_uid,
+                    MAX(cr.activity_session_completed_at) AS activity_session_completed_at,
+                    MAX(cr.extra_metadata->>'question_concept_uid') AS question_concept_uid,
+                    MAX(cr.attempt_number) AS attempt_number,
+                    BOOL_OR(cr.correct) AS correct
+                FROM concept_results cr
+                JOIN relevant_sessions act_s ON cr.activity_session_id = act_s.id
+                JOIN concept_uids ON (cr.extra_metadata->>'question_concept_uid') = concept_uids.uid
+                WHERE#{' '}
+                    /* The earliest concept_results with `question_uid` is from July 2022, so this condition limits the number of partitions we look at */
+                    cr.activity_session_completed_at BETWEEN '2022-07-01' AND NOW() + INTERVAL '1 hour'
+                    #{relevant_concept_results_where_clause}
+                GROUP BY cr.activity_session_id, cr.question_number
+        )
+      SQL
+    end
+
+    private def select_sql
```

Comment:
I mention this below in another base class, but I'd put these all below the `run` method to make the protocol clear.